### PR TITLE
[Fix-10452] Serial wait for policy recovery

### LINKED ISSUE #10452: [Feature] Serial wait for policy recovery.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

Serial wait for policy recovery.

### Use case

Serial wait for policy recovery.

### Related issues

No

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by WangJPLeo**
## Purpose of the pull request
Serial wait for policy recovery

## Brief change log

close #10452

## Verify this pull request

Manually verified the change by testing locally.

### CODE REVIEW

**Comment by SbloodyS** [comment]
Hi @WangJPLeo , there are some ut failed.

**Comment by WangJPLeo** [comment]
> 

just fixed UT, thank you @SbloodyS

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java [hunk 1]
```diff
@@ -345,25 +345,31 @@ protected void saveSerialProcess(ProcessInstance processInstance, ProcessDefinit
...
-                        try {
-                            stateEventCallbackService.sendResult(address, port, stateEventChangeCommand.convert2Command());
-                        } catch (Exception e) {
-                            logger.error("sendResultError");
-                        }
+                    );
+                    try {
+                        stateEventCallbackService.sendResult(address, port, stateEventChangeCommand.convert2Command());
+                    } catch (Exception e) {
+                        logger.error("sendResultError");
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java]:
```suggestion
                        logger.error("sendResultError", e);
```

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java [hunk 2]
```diff
@@ -345,25 +345,31 @@ protected void saveSerialProcess(ProcessInstance processInstance, ProcessDefinit
...
                             info.getId(), 0, info.getState(), info.getId(), 0
-                        );
-                        try {
-                            stateEventCallbackService.sendResult(address, port, stateEventChangeCommand.convert2Command());
-                        } catch (Exception e) {
-                            logger.error("sendResultError");
-                        }
+                    );
+                    try {
+                        stateEventCallbackService.sendResult(address, port, stateEventChangeCommand.convert2Command());
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java]: I think we should add a new `sendResult(Host host, Command command)`, WDYT?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java [hunk 3]
```diff
@@ -852,6 +851,10 @@ public void checkSerialProcess(ProcessDefinition processDefinition) {
             if (nextProcessInstance == null) {
                 return;
             }
+            ProcessInstance nextReadyStopProcessInstance = this.processService.loadNextProcess4Serial(processInstance.getProcessDefinition().getCode(), ExecutionStatus.READY_STOP.getCode(), processInstance.getId());
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java:854]: Is the state `READY_STOP` right?

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java [hunk 4]
```diff
@@ -345,25 +345,31 @@ protected void saveSerialProcess(ProcessInstance processInstance, ProcessDefinit
...
+                    continue;
+                }
+                info.setCommandType(CommandType.STOP);
+                info.addHistoryCmd(CommandType.STOP);
+                info.setState(ExecutionStatus.READY_STOP);
+                int update = updateProcessInstance(info);
+                // determine whether the process is normal
+                if (update > 0) {
+                    String host = info.getHost();
+                    String address = host.split(":")[0];
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java]: use Constant.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java [hunk 5]
```diff
@@ -345,25 +345,31 @@ protected void saveSerialProcess(ProcessInstance processInstance, ProcessDefinit
...
+                    int port = Integer.parseInt(host.split(":")[1]);
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java]: same here.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java (hunk 1 above)
Comment (WangJPLeo) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java]: ok

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java (hunk 2 above)
Comment (WangJPLeo) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java]: this is a good idea

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java (hunk 4 above)
Comment (WangJPLeo) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java]: ok thank you

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java (hunk 3 above)
Comment (WangJPLeo) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java:854]: Yes, this happens when the serial-first workflow is executed at high frequency. If there is an instance in REAY_STOP, it is not possible to wake up the instance that should be executed first.

**Review by caishunfeng — APPROVED** [review]
LGTM

**Pull request merged**